core/color.js

```diff
@@ -0,0 +1,432 @@
+/**
+ * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+/**
+ * @fileoverview Defines the {@link CKEDITOR.tools.style.color} normalizer class
+ * that parse color string code other formats.
+ */
+
+( function() {
+	'use strict';
+
+	CKEDITOR.tools.style.color = function( colorCode ) {
+		this.originalColorCode = colorCode;
+		this.parseInput( colorCode );
+	};
+
+	CKEDITOR.tools.style.color.prototype = {
+		originalColorCode: '',
+		hexColorCode: '',
+		defaultHexColorCode: '#000000',
```

Comment:
Not sure if we need this :thinking: What would be the reason/case for returning default value if invalid color is passed? And even though if we need something like thath somewhere in the code it shouldn't be color class responsibility I suppose :thinking:

Reply:
I should write it before... There are my drivers:
- whenever we **expect** User input any color string - we could almost blindly(maybe after sanitization) pass it to `color` class.
- whenever we need pass color to any method - we pass `color` object. So there will be no confusion about what kind of `string` this may be. Should we sanitize it? Maybe validate it? Instead of checking a long chain of calls before our method (to find out what is the origin of this color) - we can just use `color` class interface.
- whenever we need in end-cases pass specific string (like style property value for `background-color`) - then we use proper formatter for that, like `getHex()`. It's very clear in that case what kind of string we expect in that case on output.

Default color:
- Based on previous assumptions - if `color` class **always** provide **valid** color code - there is no place for mistakes for programmers. It always gives us a valid color, so there will be no `background-color: sthWeird`.
- The only source of untrusty (as always) is User input. Or code where we create an object "from hand" - but for that case we have tests - right?
- It may be not the expected color when User misspells `rde` instead of `red` but it's always **valid**. Like in color history case (#4333). No matter of user input we display something. But for now it's something invalid. Maybe we should check input in color palette first - then if color is invalid, just drop it - that is right. But again - in case we miss such validation on any level - there is always **valid** color code at the end.

From the other side, I can imagine that instead of default color we could throw exception during `color` creation, and other parts of code will be responsible to create a valid `color` object in that case.

So from my point of view, I try to make it more like static typed. Even - to get rid of checking if a variable is a valid object or some null. There is in "Code refactor" by Martin Fowler idea about 'default` object instead of null returning. Which makes it another driver for me.